### PR TITLE
SpannableTextView widget

### PR DESCRIPTION
**Pull request by Josvanegmond**
Allows parts of a TextView to have different styles and custom fonts.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -3,7 +3,7 @@ buildscript {
         mavenCentral()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:0.8.+'
+        classpath 'com.android.tools.build:gradle:1.2.3'
```
Comment (dzolnai) [inline build.gradle]: 1.3.0 is available :)

File: build.gradle [hunk 2]
```diff
@@ -16,7 +16,7 @@ dependencies {
 
 android {
     compileSdkVersion 17
-    buildToolsVersion "19.0.1"
+    buildToolsVersion "19.1.0"
```
Comment (dzolnai) [inline build.gradle]: latest is 21.1.2 I think

File: src/com/egeniq/utils/loader/TypefaceLoader.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.util.Hashtable;
+
+/**
+ * Created by josvanegmond on 7/27/15.
+ *
+ * Class caching Typefaces to reduce loading time
+ * Original code from https://code.google.com/p/android/issues/detail?id=9904
+ */
+public class TypefaceLoader {
```
Comment (dzolnai) [inline src/com/egeniq/utils/loader/TypefaceLoader.java]: Isn't this class the same as: https://github.com/egeniq/egeniq-android/blob/master/src/com/egeniq/utils/FontLoader.java ?

File: src/com/egeniq/widget/CustomTypefaceSpan.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import android.graphics.Typeface;
+import android.text.TextPaint;
+import android.text.style.MetricAffectingSpan;
+
+/**
+ * Created by josvanegmond on 7/27/15.
+ * Assisting class for multiple typefaces
+ * Original code from http://stackoverflow.com/a/17961854
+ */
+public class CustomTypefaceSpan extends MetricAffectingSpan {
```
Comment (dzolnai) [inline src/com/egeniq/widget/CustomTypefaceSpan.java]: This is exactly the source code of TypefaceSpan (http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/5.1.1_r1/android/text/style/TypefaceSpan.java#TypefaceSpan). Maybe you can remove this class and use the built-in one?

File: src/com/egeniq/widget/SpannableTextView.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+/**
+ * Created by josvanegmond on 7/27/15.
+ * Generic class for marking up a TextView with multiple MetricAffectingSpan objects.
+ */
+public class SpannableTextView extends TextView {
+
+    /**
+     * Data class for storing Markup data for each span applied to this TextView.
+     */
+    private class _MarkupData {
```
Comment (dzolnai) [inline src/com/egeniq/widget/SpannableTextView.java]: we don't use underscore for private class names. You can name it as you would public classes.

Comment (dzolnai) [inline src/com/egeniq/widget/SpannableTextView.java]: Also I think this class can be static

File: src/com/egeniq/widget/CustomTypefaceSpan.java (hunk 4 above)
Comment (Josvanegmond) [inline src/com/egeniq/widget/CustomTypefaceSpan.java]: It takes different constructor arguments. The built-in doesn't take a custom Typeface. This class has the custom Typeface as attribute, not the family string. I don't think it's parcelable then?

File: src/com/egeniq/utils/loader/TypefaceLoader.java (hunk 3 above)
Comment (Josvanegmond) [inline src/com/egeniq/utils/loader/TypefaceLoader.java]: It appears so. Didn't see the View independent getFont method.

File: src/com/egeniq/widget/SpannableTextView.java (hunk 5 above)
Comment (Josvanegmond) [inline src/com/egeniq/widget/SpannableTextView.java]: Isn't MarkupData associated with an instance of SpannableTextView instead of the class?

**Comment by Josvanegmond** [comment]
Updated

**Pull request merged**